08openmp/index.md

```diff
@@ -0,0 +1,307 @@
+
+I should point out here that this performance is when using the default compiler optimisation level `-O0`. If we switch to using `-O2` with `g++ -fopenmp -O2 main.cpp timer.cpp`, the serial performance drops to $0.74$s and the parallel performance to $0.10$, now giving a speedup of $740$%. While parallelisation can improve our overall performance greatly, single-core optimisations are still crucial to achieving the best performance possible from a code.
+
+### Example 2: Summing an array (naively)
+
+Let's now sum the array from the previous example. I won't go into the maths but the sum of this actually gives us a numerical approximation of the integral of $sin(\pi x)$ from $0$ to $\pi$. If we sum the array, multiply it by $\pi$ and divide by $N$ we should get *exactly* $2$. The filling of the vector is the same so I'll only print here the code for the sum itself:
+
+```
+  double sum = 0.;
+
+  Timer timer;
+
+#pragma omp parallel for
+  for(int i=0; i<vec.size(); ++i) {
+    sum += vec[i];
+  }
+
+  double elapsed = timer.elapsed();
+
+  std::cout << "Time: " << elapsed << '\n';
+  std::cout << "Result: " << M_PI*sum/N << '\n';
+```
+
+Compiling and running without OpenMP gives:
+
+```
+Time: 0.276116
+Result: 2
+```
+
+This is exactly what we expect. If you enable OpenMP however, you'll probably get something like:
+
+```
+Time: 0.957646
+Result: 0.14947
+```
+
+I recommend you take a break from reading/coding here and think about the questions, *why has the time increased with more threads* and *why is the answer wrong*? Don't worry if it's not obvious, this is a subtle and tricky bug and requires thinking about what each thread is doing as it accesses each variable inside the loop.
+
+---
+
+This behaviour is the result of a **data race**. Consider even just two threads working in parallel inside the loop. Thread 1 reads data from `vec[i]`. The `for` construct has split the loop between the two threads so its value of `i` is totally unique to thread 1 and it is the only thread touching the data in `vec[i]`. It then needs to add this value into `sum`. It reads `sum`, adds `vec[i]` to it and writes it back into `sum`. This addition takes some time however, and thread 2 is performing the exact same operation but using a different value of `i`. When thread 1 reads `sum` and is currently spending time adding, thread 2 could easily have just finished an addition and has written a new value to `sum`. Thread 1 has no idea that thread 2 has done this and then unknowingly *overwrites* the updated value of `sum` with its own version. Not only does this mean the value of `sum` is *incorrect* but threads are not allowed to write to the same variable at the same time, so each threads must wait for the other to finish, adding more time to the overall calculation. This is why the overall time has increased, despite using multiple threads, and why the result is wrong!
```

Comment:
```suggestion
This behaviour is the result of a **data race**. Consider even just two threads working in parallel inside the loop. Thread 1 reads data from `vec[i]`. The `for` construct has split the loop between the two threads so its value of `i` is totally unique to thread 1 and it is the only thread touching the data in `vec[i]`. It then needs to add this value into `sum`. It reads `sum`, adds `vec[i]` to it and writes it back into `sum`. This addition takes some time however, and thread 2 is performing the exact same operation but using a different value of `i`. While thread 1 is occupied performing the addition, thread 2 could finish an addition operation of its own and write a new value to `sum`. Thread 1 has no idea that thread 2 has done this and then unknowingly *overwrites* the updated value of `sum` with its own version, thereby removing the contribution to the sum calculated by thread 2. Not only does this mean the value of `sum` is *incorrect* but threads are not allowed to write to the same variable at the same time, so each thread must wait for the other to finish, adding more time to the overall calculation. This is why the overall time has increased, despite using multiple threads, and why the result is wrong!
```
This wording is just a little clearer to me but totally up to you! 

In the longer term I think this would be great with some step by step diagrams to show how this works because it can be hard to visualise in your head if you're not very familiar with the way that CPUs work.